src/oauth2/client.cr

```diff
@@ -54,18 +54,35 @@
 # You can also use an `OAuth2::Session` to automatically refresh expired
 # tokens before each request.
 class OAuth2::Client
+  private getter host, client_id, client_secret, port, scheme, authorize_uri,
+    token_uri, redirect_uri, auth_scheme
+
+  # returns the HTTP::Client instance with `basic_auth` set if appropriate
+  private getter http_client : HTTP::Client do
+    HTTP::Client.new(URI.new(@scheme, @host, @port)).tap do |client|
+      client.basic_auth(client_id, client_secret) if auth_scheme.http_basic?
+    end
+  end
+
   # Creates an OAuth client.
   #
   # Any or all of the customizable URIs *authorize_uri* and
   # *token_uri* can be relative or absolute.
```

Comment:
So this was what I was talking about in https://github.com/crystal-lang/crystal/pull/9127#discussion_r410750791 with "complex handling".

This implementation will break if you pass an absolute URL as `token_uri` now. I don't even know what will happen if you pass an absolute URL, possibly for a different host, to an instance of `HTTP::Client`, but I can't imagine it to just work.

Now the question is, do we want to keep supporting this or do we say the token endpoint being on another host is just too uncommon? I would be fine with either.

In case we keep it like this, the docs here need to be updated and this needs to be documented in the changelog as a breaking change, even more so. In case we go back I'd still be okay with the reimplementation of `HTTP::Client#basic_auth` in here.

What do others think?

Reply:
Actually I tested that it worked OK with an absolute URL as token URI:

```
require "oauth2"
require "base64"

host = ENV["HOST"]
client_id = ENV["CLIENT_ID"]
client_secret = ENV["CLIENT_SECRET"]
token_uri_suffix = ENV["TOKEN_URI_SUFFIX"]

token_uri = "https://#{host}#{token_uri_suffix}"

oauth2_client = OAuth2::Client.new(
  host: host,
  client_id: client_id,
  client_secret: client_secret,
  token_uri: token_uri
)

begin
  access_token = oauth2_client.get_access_token_using_client_credentials
  puts access_token.access_token
  puts access_token.extra
rescue ex : OAuth2::Error
  puts ex.class
  puts ex.message
end
```

```
eslater:oauth-test/ $ /Users/eslater/code/crystal/lang/crystal/bin/crystal oauth2-test.cr
J10i_m06f_wOiMS5QoTupA
```

Reply:
I'm happy to revert to the prior approach of just setting an auth header and using a one shot `HTTP::Client` if you prefer.

Reply:
Mmh, did you test what happens if the host in the token URI is different to the one given to `OAuth2::Client`?

Reply:
No I didn't, because I don't have an environment like that available.

Reply:
Maybe it's possible to write a spec for this somehow 🤔